### DISCUSSION TITLE
Could I pass some model into the template?

### DISCUSSION THREAD

**Discussion by ignatandrei** [atifaziz/Transplator#2] [Q&A] [answered]
I cannot see in the examples if I can pass some Model to the template. It is possible ?

**Comment by ignatandrei** [comment]
Or how I could interfere with outside code  ?

**Comment by atifaziz** [comment, marked as answer]
At present, I have designed two modes of operation depending on the first parsed token of the template:

1. If the first token is a text literal then a simple _partial_ class is written based on the file name plus the `Template` suffix. The class will have a single, private, instance-based and parameter-less method called `RenderCore`.
2. If the first token is code (`{%`…`%}`) then you are responsible for providing the entire class and method definition, thus having complete control over what and how you allow to be passed as a model (e.g. via one or more parameters).

In either mode, you have to provide two additional methods that must have the following signatures:

- `WriteText(string value)`: called to emit literal text tokens of the template
- `WriteValue(object value)`: called to emit the text for a run-time value

In fact, the second method can have overloads to avoid boxing. There is no requirement that the `value` parameter be typed as `object`.

An example of 1 is [the `HelloWorld.cstt` sample](https://github.com/atifaziz/Transplator/blob/ee1cb572d40b38747b275ec92f4b06cc496744d7/eg/HelloWorld.cstt):

https://github.com/atifaziz/Transplator/blob/ee1cb572d40b38747b275ec92f4b06cc496744d7/eg/HelloWorld.cstt#L1-L6

This will generate a partial class called `HelloWorldTemplate` with a `RenderCore` method.

In [the sample project](https://github.com/atifaziz/Transplator/blob/ee1cb572d40b38747b275ec92f4b06cc496744d7/eg), the partial definition is then completed in `Program.cs`:

https://github.com/atifaziz/Transplator/blob/ee1cb572d40b38747b275ec92f4b06cc496744d7/eg/Program.cs#L30-L40

It inherits from `Template` that provides implementations for `WriteText` and `WriteValue` that emit to a `StringBuilder`:

https://github.com/atifaziz/Transplator/blob/ee1cb572d40b38747b275ec92f4b06cc496744d7/eg/Program.cs#L17-L28

There is no requirement for inheritance. This is just one setup I came up with for the samples.

The usage is then a simple call to the public and static `Render` method:

https://github.com/atifaziz/Transplator/blob/ee1cb572d40b38747b275ec92f4b06cc496744d7/eg/Program.cs#L4

An example of mode 2 is [`Parameters.cstt`](https://github.com/atifaziz/Transplator/blob/ee1cb572d40b38747b275ec92f4b06cc496744d7/eg/Parameters.cstt):

https://github.com/atifaziz/Transplator/blob/ee1cb572d40b38747b275ec92f4b06cc496744d7/eg/Parameters.cstt#L1-L24

As you can see, you have to provide the complete surrounding class and rendering method. You can name and define the class and method as you like. What's more, the _rendering method_ can have whatever parameters you need. The model in the case above is a sequence of strings (`ordinals`). Also, `ParametersTemplate` does not inherit from `Template` as in the case of `HelloWorldTemplate` and provides the `WriteText` and `WriteValue` methods as local functions. It goes to show that you have complete freedom.

In the first mode of operation, you can also provide a model, but it would have to be done out-of-band with the `RenderCore` method that is generated. Suppose we change `HelloWorld.cstt` so that it refers to some `Time` _model_ instead of using the `DateTime.Now` static:

```
Hello world!

Today, the date is {% Time.ToString("MMM dd, yyyy") %}.
The current time is {% Time.ToString("HH:mm") %}.
```

Here is how then you would adapt `HelloWorldTemplate` to allow passing the model:

```c#
partial class HelloWorldTemplate : Template
{
    HelloWorldTemplate(DateTime time, StringBuilder? sb = null) : base(sb) => Time = time;

    DateTime Time { get; }

    public static string Render(DateTime time)
    {
        var template = new ModelTemplate(time);
        template.RenderCore();
        return template.ToString();
    }
}
```

The public and static `Render` method now takes the model and make it available with the `Time` property (initialized through a private constructor).

Hope this sheds some light into options for passing the model.

**Reply by ignatandrei** [reply]
Thanks!
